packages/jsii-pacmak/lib/targets/python/type-name.ts

```diff
@@ -0,0 +1,354 @@
+  private static readonly STR = new Primitive('str', {});
+  private static readonly JSON = new Primitive('typing.Mapping[typing.Any, typing.Any]', { 'typing': new Set(['']) });
+
+  public static readonly ANY = new Primitive('typing.Any', { 'typing': new Set(['']) });
+  public static readonly NONE = new Primitive('None', {});
+
+  public static of(type: PrimitiveTypeReference): TypeName {
+    switch (type.primitive) {
+      case PrimitiveType.Boolean:
+        return Primitive.BOOL;
+      case PrimitiveType.Date:
+        return Primitive.DATE;
+      case PrimitiveType.Number:
+        return Primitive.JSII_NUMBER;
+      case PrimitiveType.String:
+        return Primitive.STR;
+      case PrimitiveType.Json:
+        return Primitive.JSON;
+      case PrimitiveType.Any:
+      default:
+        return Primitive.ANY;
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  readonly #pythonType: string;
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  readonly #requiredImports: PythonImports;
+
+  private constructor(pythonType: string, requiredImports: PythonImports) {
+    this.#pythonType = pythonType;
+    this.#requiredImports = requiredImports;
+  }
+
+  public pythonType() {
+    return this.#pythonType;
+  }
+
+  public requiredImports() {
+    return this.#requiredImports;
+  }
+}
+
+class Union implements TypeName {
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  readonly #options: readonly TypeName[];
+
+  public constructor(options: readonly TypeName[]) {
+    this.#options = options;
+  }
+
+  public pythonType(context: NamingContext) {
+    return `typing.Union[${this.#options.map(o => o.pythonType(context)).join(', ')}]`;
+  }
+
+  public requiredImports(context: NamingContext) {
+    return mergePythonImports(
+      ...this.#options.map(o => o.requiredImports(context)),
+      { typing: new Set(['']) },
+    );
+  }
+}
+
+class UserType implements TypeName {
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  readonly #fqn: string;
+
+  public constructor(fqn: string) {
+    this.#fqn = fqn;
+  }
+
+  public pythonType(context: NamingContext) {
+    return this.resolve(context).pythonType;
+  }
+
+  public requiredImports(context: NamingContext) {
+    const requiredImport = this.resolve(context).requiredImport;
+    if (requiredImport == null) {
+      return {};
+    }
+    return { [requiredImport.sourcePackage]: new Set([requiredImport.item]) };
+  }
+
+  private resolve({ assembly, submodule, nestingScope, typeAnnotation = true }: NamingContext) {
+    const { assemblyName, pythonFqn } = toPythonFqn(this.#fqn, assembly);
+    if (assemblyName !== assembly.name) {
+      return {
+        pythonType: pythonFqn,
+        requiredImport: {
+          // TODO: Migth we need to import a subpackage instead?
+          sourcePackage: toPythonFqn(assemblyName, assembly).pythonFqn,
```

Comment:
do we have a test that imports across from a subpackage?

Reply:
Oh I had forgotten this TODO. Definitely need to add one there!